services/auth/middleware.go

```diff
@@ -235,6 +236,8 @@ func VerifyAuthWithOptionsAPI(options *VerifyOptions) func(ctx *context.APIConte
 					return
 				}
 				if !ok {
+					audit.Record(audit.UserAuthenticationFailTwoFactor, ctx.Doer, ctx.Doer, twofa, "Failed two-factor authentication for user %s.", ctx.Doer.Name)
```

Comment:
There are some sensitive fields in the `twofa` struct.

Is it right to write the sensitive data (eg: secret, salt, hash) into a file?

The same question also applies to other objects, like user or others.